packages/docusaurus-utils/src/index.ts

```diff
@@ -252,43 +252,77 @@ export function createExcerpt(fileString: string): string | undefined {
 }
 
 type ParsedMarkdown = {
-  frontMatter: {
-    // Returned by gray-matter
-    // eslint-disable-next-line @typescript-eslint/no-explicit-any
-    [key: string]: any;
-  };
+  // Returned by gray-matter
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  frontMatter: Record<string, any>;
   content: string;
   excerpt: string | undefined;
+  hasFrontMatter: boolean;
 };
-export function parseMarkdownString(markdownString: string): ParsedMarkdown {
-  const options: Record<string, unknown> = {
-    excerpt: (file: matter.GrayMatterFile<string>): void => {
-      // Hacky way of stripping out import statements from the excerpt
-      // TODO: Find a better way to do so, possibly by compiling the Markdown content,
-      // stripping out HTML tags and obtaining the first line.
-      file.excerpt = createExcerpt(file.content);
-    },
-  };
 
+export function readFrontMatter(
+  markdownString: string,
+  source?: string,
+  options: Record<string, unknown> = {},
+  removeTitleHeading = true,
+): ParsedMarkdown {
   try {
-    const {data: frontMatter, content, excerpt} = matter(
-      markdownString,
-      options,
-    );
-    return {frontMatter, content, excerpt};
+    const result = matter(markdownString, options);
+    result.data = result.data || {};
+    result.content = result.content.trim();
+
+    const hasFrontMatter = Object.keys(result.data).length > 0;
+
+    const heading = /^# (.*)[\n\r]?/gi.exec(result.content);
+    if (heading) {
+      if (result.data.title) {
+        console.warn(
+          `Duplicate title detected in \`${source || 'this'}\` file`,
+        );
+      } else {
+        result.data.title = heading[1].trim();
+        if (removeTitleHeading) {
+          result.content = result.content.replace(heading[0], '');
+          if (result.excerpt) {
+            result.excerpt = result.excerpt.replace(heading[1], '');
+          }
```

Comment:
wonder if we could extract this easily in a separate function and test this in isolation?

Reply:
we could actually go few steps further in separate pr and have index only with few `export * from '...'` and split utils to mutiple files that does specific things as currently this file is starting to become really long and unreadable

Reply:
totally agree :)

was like that before and want to refactor that for a long time already